src/reconciler/hotReplacementRender.js

```diff
@@ -129,17 +129,19 @@ const mapChildren = (children, instances) => ({
       }
     }
 
-    const newChildren = asArray((child.props && child.props.children) || [])
+    const newChildren = asArray(
+      (child.props && child.props.children) || child.children || [],
```

Comment:
When result of this function on next render iteration got in here again - it will be just ignored.
That was the first problem.